drivers/interrupt_controller/intc_miwu.c

```diff
@@ -0,0 +1,368 @@
+ *
+ * This driver uses device tree files to present the relationship bewteen
+ * MIWU and the other devices in different npcx series. For npcx7 series,
+ * it include:
+ *  1. npcxn-miwus-wui-map.dtsi: it presents relationship between wake-up inputs
+ *     (WUI) and its source device such as gpio, timer, eSPI VWs and so on.
+ *  2. npcx7-miwus-int-map.dtsi: it presents relationship between MIWU group
+ *     and NVIC interrupt in npcx7. Please notice it isn't 1-to-1 mapping.
+ *     For example, here is the mapping between miwu0's group a & d and IRQ7:
+ *
+ *     map_miwu0_groups: {
+ *         parent = <&miwu0>;
+ *         group_ad0: group_ad0_map {
+ *             irq        = <7>;
+ *             group_mask = <0x09>;
+ *         };
+ *         ...
+ *     };
+ *
+ *     It will connect IRQ 7 and intc_miwu_isr0() with the argument, group_mask,
+ *     by IRQ_CONNECT() during driver initialization function. With group_mask,
+ *     0x09, the driver checks the pending bits of group a and group d in ISR.
+ *     Then it will execute related callback functions if they have been
+ *     registered properly.
+ *
+ * INCLUDE FILES: soc_miwu.h
+ *
+ */
+
+#include <assert.h>
+#include <device.h>
+#include <kernel.h>
+#include <soc.h>
+#include <irq_nextlevel.h>
+#include <drivers/gpio.h>
+
+#include "soc_miwu.h"
+#include "soc_gpio.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(intc_miwu, LOG_LEVEL_ERR);
+
+/* MIWU module instances forward declaration */
+static struct device *miwu_devs[];
+
+/* Driver config */
+struct intc_miwu_config {
+	/* miwu controller base address */
+	uint32_t base;
+	/* index of miwu controller */
+	uint8_t index;
+};
+
+/* Callback functions list for GPIO wake-up inputs */
+sys_slist_t cb_list_gpio;
+
+/*
+ * Callback functions list for the generic hardware modules  wake-up inputs
+ * such as timer, uart, i2c, host interface and so on.
+ */
+sys_slist_t cb_list_generic;
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct intc_miwu_config *)(dev)->config)
+
+BUILD_ASSERT(sizeof(struct miwu_io_callback) == sizeof(struct gpio_callback),
+	"Size of struct miwu_io_callback must equal to struct gpio_callback");
+
+BUILD_ASSERT(sizeof(struct miwu_io_params) == sizeof(gpio_port_pins_t),
+	"Size of struct miwu_io_params must equal to struct gpio_port_pins_t");
+
+/* MIWU local functions */
+static void intc_miwu_dispatch_gpio_isr(uint8_t wui_table,
+					uint8_t wui_group, uint8_t wui_bit)
+{
+	struct miwu_io_callback *cb, *tmp;
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&cb_list_gpio, cb, tmp, node) {
+		/* Pending bit of wui group match the wui item in list */
+		if (cb->params.wui.table == wui_table
+			&& cb->params.wui.group == wui_group
+			&& cb->params.wui.bit == wui_bit) {
+			__ASSERT(cb->handler, "No GPIO callback handler!");
+			/* Execute GPIO callback function */
+			cb->handler(soc_get_gpio_dev(cb->params.gpio_port),
+					(struct gpio_callback *)cb,
+					cb->params.pin_mask);
+		}
```

Comment:
Do you expect more than one call back to match? If not, let's break. If so, let's add a comment saying that multiple GPIO callback may match

Reply:
Done. I think it is not one to one relationship and add more comment here.